### ISSUE TITLE
Extend schema to allow entries in multiple categories

### ISSUE THREAD

**Issue by jwishnie** [2factorauth/twofactorauth#3707]
Many service providers offer a _suite_ of services behind a single login.

e.g. Google and Microsoft accounts provide email, chat, backup, etc...

Rather than recreating entries for each service in each separate category, we could extend the schema for entries to contain a list of categories in which the entry should appear.

**Comment by kmpoppe** [comment]
Second that.

**Comment by mxxcon** [comment]
This was previously discussed already, but nobody stepped up with ideas on the shcema for this and other implementation intricacies.

**Comment by jakejarvis** [comment]
I literally just had the same thought a few minutes ago, too – currently, if Microsoft suddenly announced support for hardware tokens, it would require a change to _nine_ files!

But agreed that this would be a pretty major undertaking... This would mean a complete restructuring of the `_data` directory from category-based to service-based, to the point where each service would need its own file and its category/categories listed within it, no?

**Comment by pm17788** [comment]
This almost feels like each "entry" (say, Microsoft as an example) should simply have multiple tags indicating categories, and either at load time or some sort of a macro thing would generate data.
JSON-ish view, I suppose:
```json
{
  "Companies": {
    "Microsoft": {
      "Services": [
        "email",
        "backups",
        "productivity"
      ],
      "2FA Methods":[
        "SMS",
        "HW Token",
        "SW Token",
        "Scratch-n-Sniff"
      ],
      "URL": "https://www.microsoft.com",
      "PrettyLogoPicturePath": "images/Microsoft/"
    }
  }
}
```

So, an entry like this would "show up" under `emails`, `backups`, and `productivity`, but it's one entry, with the common dataset.

I'm _certain_ I'm missing something in the intricacies, and would be glad to bounce ideas back and forth.

**Comment by jwishnie** [comment]
@pm17788 that's exactly the sort of thing I was thinking about. Essentially instead of separate yml's for each category, there's a single list of login services ('identity providers') and each entry lists all the services it provides access to...

@jakejarvis MS _has_ announced they will be supporting security keys visa Windows Hello, and _currently_ support keys on personal MS accounts...

**Comment by pm17788** [comment]
@jwishnie: So, what's our next step on this? :) How can I help move this forward?

**Comment by kmpoppe** [comment]
Coming back to this.

Combining what @jwishnie and @pm17788 said, I'd go for:

1. A new "invisible" category (e.g. `Identity providers`), that just says which services it provides (sms, software, ...) including documentation.
2. Data-pages get a new field (e.g. `identity_provider`) (which can be a list, as this is a one-to-many-connection)
3. The "Docs" column gets a fancy, new icon (for identity provider), that, just like the exception box, pops up and says which i-prov support 2fa for that site.
4. The 2FA options are compiled to show those, that are supported by at least 1 out of the i-provs or the page directly.

IMHO, making the connection from the site-entry in the data-ymls towards the identity-provider and not vice-versa, keeps up the idea of having every information relating to a given entry in one place (reads: one file). I'm totally open for any other interpretation of this idea.

This means a considerable amount of additional programming and I am unsure (due to lack of experience with ruby) how exactly this would be accomplished.

**Comment by kmpoppe** [comment]
While working on another topic I found that

* Rackspace
* UpCloud
* VentraIP

Are the exact same entries in Hosting/VPS and Domains, with the first two pointing to the same documentation, yet they show different 2FA-options, which is obviously wrong.
This issue could be addressed as well.

**Comment by dsoegijono** [comment]
If we list a site on different categories, when we use the search function, the same site would be displayed over and over again in each category right? Do we also want to rethink the search function then?

**Comment by pm17788** [comment]
> If we list a site on different categories, when we use the search function, the same site would be displayed over and over again in each category right? Do we also want to rethink the search function then?

(Apologies if I didn't understand the implication of what you're saying) Same site showing up under multiple categories doesn't _seem_ that horrible.

**Comment by hon1nbo** [comment]
Sorry for the Necro but, having heard from people I've referred to this site over the years, this is the biggest negative feedback I hear about it. Users aren't just looking for a service they already know the name of, but for new services within a vertical that support strong authentication.

If I'm reading this right, it seems that just no one has stepped up to write a viable implementation to fix the quirks; is that correct? There is a proposed schema in a comment above but no implementation for it.

Is there any opposition if someone has the time to write an implementation for that schema and a parser that can take the existing files and sort them into it? I looked through issues and gitter for more discussion on the matter but it seems like there isn't too much comprehensive discussion I can find outside of this thread (gitter isn't easy to search as-is from an initial look). I've got time on my hands now and I reference this site with people often enough to justify spending some of it here.
